cluster-autoscaler/core/equivalence_groups.go

```diff
@@ -51,39 +53,54 @@ type equivalenceGroup struct {
 	representant *apiv1.Pod
 }
 
+const maxEquivalenceGroupsByController = 10
+
 // groupPodsBySchedulingProperties groups pods based on scheduling properties. Group ID is meaningless.
+// TODO(x13n): refactor this to have shared logic with PodSchedulableMap.
 func groupPodsBySchedulingProperties(pods []*apiv1.Pod) map[equivalenceGroupId][]*apiv1.Pod {
 	podEquivalenceGroups := map[equivalenceGroupId][]*apiv1.Pod{}
 	equivalenceGroupsByController := make(map[types.UID][]equivalenceGroup)
 
 	var nextGroupId equivalenceGroupId
 	for _, pod := range pods {
 		controllerRef := drain.ControllerRef(pod)
-		if controllerRef == nil {
+		if controllerRef == nil || pod_utils.IsDaemonSetPod(pod) {
 			podEquivalenceGroups[nextGroupId] = []*apiv1.Pod{pod}
 			nextGroupId++
 			continue
 		}
 
-		matchingFound := false
-		for _, g := range equivalenceGroupsByController[controllerRef.UID] {
-			if reflect.DeepEqual(pod.Labels, g.representant.Labels) && utils.PodSpecSemanticallyEqual(pod.Spec, g.representant.Spec) {
-				matchingFound = true
-				podEquivalenceGroups[g.id] = append(podEquivalenceGroups[g.id], pod)
-				break
-			}
+		egs := equivalenceGroupsByController[controllerRef.UID]
+		if gid := match(egs, pod); gid != nil {
+			podEquivalenceGroups[*gid] = append(podEquivalenceGroups[*gid], pod)
+			continue
 		}
-
-		if !matchingFound {
-			newGroup := equivalenceGroup{
-				id:           nextGroupId,
-				representant: pod,
-			}
-			equivalenceGroupsByController[controllerRef.UID] = append(equivalenceGroupsByController[controllerRef.UID], newGroup)
-			podEquivalenceGroups[newGroup.id] = append(podEquivalenceGroups[newGroup.id], pod)
-			nextGroupId++
+		newGroup := equivalenceGroup{
+			id:           nextGroupId,
+			representant: pod,
 		}
+		equivalenceGroupsByController[controllerRef.UID] = append(egs, newGroup)
```

Comment:
I think the current implementation works, but it's surprising to me that we use the maxEquivalenceGroupsByController to limit the lookup through equivalence groups list, not to limit the size of the list itself. I think the code would be easier to understand if we just added a more explicit check + continue, rather than keep extending the list and bypass comparison like we do now.

Also I highly doubt it makes any real performance difference, but it feels pointless to create all those equivalenceGroup structs if we're never going to return them or look into them.